### PR TITLE
API - Improve ACL processing, deprecate `hook_civicrm_notePrivacy`

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Expands the capabilities of ACL checking for API get actions while improving efficiency.

Includes documentation: https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/1118

Before
----------------------------------------
- ACL query in APIs are limited to AND operator and only one field at a time.
- Large slow subqueries were being added to every `entity_table`/`entity_id` combo.
- `hook_civicrm_notePrivacy` was supported by the smarty/quickform layer only, APIv3/APIv4/SearchKit never called it.

After
----------------------------------------
- API ACL queries can use OR operator and reference multiple fields on the entity, this is especially helpful for `entity_table`/`entity_id` combos.
- The WHERE and ON clause is used to optimize the subquery so that e.g. a search for Contact Notes will only generate ACL queries for Contacts, not every possible entity the Note could be for, since it's already being filtered to exclude those entities.
- `hook_civicrm_notePrivacy` noisily deprecated but *also* grandfathered into APIv4 so that it is called from SearchKit. This will provide 1-1 feature parity between the old Contact summary "Notes" tab and a new SearchKit-driven tab.

Technical Details
----------------------------------------
`hook_civicrm_notePrivacy` was nice [when it was added](https://civicrm.org/blogs/allenshaw/adding-privacy-and-comments-civicrm-notes) but (no disrespect to @twomice) is no longer a good way to enforce ACLs. It's designed to be triggered once for every row returned by the query, which is horribly inefficient. Also it only works for one entity (notes) and it never worked with the API/SearchKit.
This PR adds a transitional grace period where the hook is now called by APIv4/SearchKit, but emits a noisy deprecation notice if used, alerting people to switch their hook code to the newer `hook_civicrm_selectWhereClause`, which works for all entities.

Notes
----------
This changes the signature for BAO function `addSelectWhereClause()`. I did a `universe` search and it doesn't appear to be overridden outside of core, but if it is, better to change it now while the unforgiving PHP8 isn't as widely adopted.

### CODE REVIEW

**Comment by colemanw** [comment]
retest this please

**Comment by colemanw** [comment]
Wonky test failure wasn't exactly *caused* by this PR, but somehow this PR managed to piss off an already ornery test. Fixed via https://github.com/civicrm/civicrm-core/pull/27520

File: CRM/Core/DAO.php [hunk 1]
```diff
@@ -3117,25 +3157,47 @@ protected static function getDynamicFkAclClauses(): array {
...
-    }
-    $clauses = [];
-    foreach ((array) $bao->addSelectWhereClause() as $field => $vals) {
-      $clauses[$field] = NULL;
-      if ($vals) {
-        $clauses[$field] = "(`$tableAlias`.`$field` " . implode(" AND `$tableAlias`.`$field` ", (array) $vals) . ')';
-        if (empty(static::getSupportedFields()[$field]['required'])) {
-          $clauses[$field] = "(`$tableAlias`.`$field` IS NULL OR {$clauses[$field]})";
+    $tableAlias = $tableAlias ?? $bao->tableName();
+    $entityName = $entityName ?? CRM_Core_DAO_AllCoreTables::getBriefName($bao::class);
```
Comment (totten) [inline CRM/Core/DAO.php:3173]: @colemanw I was just doing a local upgrade from 5.66@head to master@head. This failed:

```
$ cv upgrade:db -vv

Box Requirements Checker
========================

> Using PHP 7.4.29
> PHP is using the following php.ini file:
  WARNING: No configuration file (php.ini) used by PHP!

> Checking Box requirements:
  .....


 [OK] Your system is ready to run the application.



Fatal error: Cannot use ::class with dynamic class name in /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/CRM/Core/DAO.php on line 3173
```

Comment (colemanw) [reply CRM/Core/DAO.php:3173]: @totten is it a php version thing? Would it work better with `get_class()`?

Comment (totten) [reply CRM/Core/DAO.php:3173]: @colemanw Yeah, I think that's right. `get_class()` seemed to work locally. I encountered two spots with the issue. Sent a follow-up PR for those two. (27536)

**Pull request merged**